Add gram-schmidt fast approximation tangent generation algorithm

aloucks opened the pull request:
# Provide a fast alternative algorithm for tangent generation

Fixes: #17834
Related (maybe): #17877

## Solution

Add a new enum `TangentStrategy` that gives the user the option to select how tangents are generated. 

The GLTF loader now has a property to set which algorithm is used for models that are missing tangents.

The `FastApproximation` strategy is implemented with the Gram-Schmidt fast approximation technique. The `HighQuality` variant uses MikkTSpace.

## Testing

Unit tests have been added to assert that both tangent computation strategies produce similar tangents for simple shapes. I also tried to check several examples to verify that the visual results are acceptable with both algorithms. I wasn't able to tell any visible difference.

> How can other people (reviewers) test your changes? Is there anything specific they need to know?

I checked the `deferred_rendering`, `parallax_mapping`, and `anisotropy` examples. Suspiciously, when I produced intentionally invalid tangents, the deferred_rendering and parallax_mapping examples seemed to render fine. I didn't try the anisotropy example with intentionally invalid values. I'm curious if there is an unrelated bug in the shader where tangents are ignored (or the examples I tested somehow didn't leverage tangents). 

## Migration Guide

- `GenerateTangentsError::MikktspaceError` has been renamed `GenerateTangentsError::AlgorithmError`
- The `Mesh::generate_tangents` and `Mesh::with_generated_tangents` methods are now deprecated. They are replaced by `compute_tangents`  and `with_computed_tangents`, both of which accept a `TangentStrategy` parameter. The naming conventino of these methods now matches `compute_normals`.

aloucks commented on crates/bevy_gltf/src/lib.rs:
```diff
@@ -124,6 +124,8 @@ pub mod prelude {
 pub struct GltfPlugin {
     custom_vertex_attributes: HashMap<Box<str>, MeshVertexAttribute>,
+    /// The strategy to use when computing mesh tangents.
+    pub computed_tangent_strategy: TangentStrategy,
```
renamed

aloucks committed:
Remove unnecessary Vec4 to [f32; 4] conversion

aloucks committed:
Fix compute_tangents doc links in Mesh::ATTRIBUTE_TANGENT

hukasu commented:
the `anisotropy` example does use tangents, this issue shows some issues caused by not optimal tangents #16179

hukasu commented:
i was making tangent generation for the primitives, but with a very naive algorithm #17691, if your method works without causing seams then my PR becames redundant

aloucks commented:
> the `anisotropy` example does use tangents, this issue shows some issues caused by not optimal tangents #16179

The sphere looks much better with the gram-schmidt tangents. I actually left `Sphere` out of the tests that I added because there was at least one vertex where the mikktspace generated tangent differed from the gram-schmidt tangent enough that it was outside of the test threshold.

## mikktspace

https://github.com/user-attachments/assets/da7e29eb-8169-4696-b402-51c5cf435967

## gram-schmidt

https://github.com/user-attachments/assets/763b3bea-de11-4ff6-b6cf-dfc1df21e626

aloucks commented:
> Can the strategy names be changed to just the name of the algorithm? Mikktspace is very dependant on the UVs, so just calling it `HighQuality` might be misleading

Bikeshedding time! 🎨 

@JMS55 had suggested to inject `Calculation` into the name, but I think it's less necessary if we change `Strategy` to `Algorithm`.  The word "algorithm" implies the "calculation" part, imo. Also, do we want to keep HighQuality/Mikktspace the default?

@JMS55  / @hukasu  what are your thoughts?

```rust
pub enum TangentAlgorithm {
  Mikktspace,
  GramSchmidt,
}
```

aloucks committed:
Fix merge conflict

aloucks committed:
Fixing merge conflict again

aloucks commented:
@JMS55 Anymore feedback on the enum and variant naming? Do we want to leave mikktspace the default or change to gram-schmidt?

IQuick143 commented on crates/bevy_mesh/src/gramschmidt.rs:
```diff
@@ -0,0 +1,216 @@
+
+            for i in 0..vertex_count {
+                let normal = Vec3A::from_array(normals[i]);
+                let tangent = tangents[i].normalize();
```
Is there a numerical reason to normalize here?
The next line already involves a normalization and is linear in the `tangent` variable.

On the other hand, I don't see the normal getting normalised, and that would cause problems in the G-S algo.

aloucks replied:
> Is there a numerical reason to normalize here?
> The next line already involves a normalization and is linear in the tangent variable.

The examples I've seen either normalize the accumulated tangent it or average it by dividing by the number of vertices that were used to accumulate it. I don't think the direction would be correct if we only normalized the final result and omitted the normalizing here.

> On the other hand, I don't see the normal getting normalised, and that would cause problems in the G-S algo.

This assumes that the vertex normals are normalized. Is that assumption false? If so, I'll need to normalize the normal here.

IQuick143 replied:
I'm not familiar with the code enough to determine whether the normals are normalised, although I would expect so.

Regarding the pre-normalisation of the tangent, that part is mathematically unnecessary as the projection is linear.
```
tangent = P(t).normalise()
```
where `P` is the projection operator `x -> x - x.dot(normal) * normal`, which is linear.

This tells us that if `t` is not normalised by a positive factor `k` we have:
```
P(k*t).normalise() = (k * P(t)).normalise() = P(t).normalise()
```
Where the first inequality comes from the linearity of P, and the second from normalisation destroying these scalar factors.

So from a purely theoretical standpoint the normalisation has no effect.

The only reason I can think of for the normalisation would be to avoid some sort of numerical error, although I cannot really think of the reasons it would be relevant here given how floating point values work.

aloucks replied:
@IQuick143 

How about this? 

```rust
let normal = Vec3A::from_array(normals[i]);
// Assume that the normal has not been pre-normalized
let normal = normal.normalize();
// Although it is not mathematically necessary to normalize the accumulated
// tangent, doing so may provide better numerical stability in extreme cases.
let tangent = tangents[i].normalize();
// Gram-Schmidt orthogonalization
let tangent = (tangent - normal * normal.dot(tangent)).normalize();
```

IQuick143 replied:
You can save a few operations (a square root and some divides) by doing:
```rust
let tangent = (tangent - normal * (normal.dot(tangent) / normal.magnitude_squared())).normalize();
```
Instead of normalising the normal.

aloucks replied:
> You can save a few operations (a square root and some divides) by doing:
> 
> ```rust
> let tangent = (tangent - normal * (normal.dot(tangent) / normal.magnitude_squared())).normalize();
> ```
> 
> Instead of normalising the normal.

Unfortunately that causes other issues where the normal is broken and has length of zero. The `cone` unit test fails. This is why I pushed a later commit to use `normalize_or_zero` instead of `normalize` when normalizing the normal.

aloucks committed:
Ensure the normal is is normalized in Gram-Schmidt tangent computation

Also added a note about why the accumulated tangent is normalized
prior to applying Gram-Schmidt orthogonalization. Normalizing the
accumulated tangent may not be mathematically necessary, but it
may prevent numerical stability issues in some cases.

aloucks committed:
Rename TangentCalculationStrategy to TangentAlgorithm

The variants have also been renamed to match their algorithm names.

aloucks commented:
@hukasu I renamed the enum to `TangentAlgorithm` and the associated variants. Any chance this can land before `0.16` is released?

hukasu commented:
i don't have authority to say when it can be merged

aloucks commented:
@alice-i-cecile Is this too late for 0.16?

alice-i-cecile commented:
Yeah, we're down to "essential bug fixes and polish" at this stage, sorry!

aloucks commented:
@alice-i-cecile Now that 0.16 is out, could this be considered for 0.17?

ChristopherBiscardi commented:
This is pretty reasonable, especially since it is adding a new option, but would need an update to merge as it has unfortunately been left off a few milestones now and has developed conflicts.

IceSentry commented:
The conflicts need to be resolved but once they are I'm happy with the solution to support multiple tangent algorithm. I don't know about the actual math though. I have no opinions about it.
